### PR TITLE
Rework `DtoCallable` to remove `getPointerElementType`

### PR DESCRIPTION
**Pull request by thewilsonator**
I'm not at all sure about this. Any inputs welcome. (also idk if this is the best code style either)

### CODE REVIEW

File: gen/tocall.cpp [hunk 1]
```diff
@@ -874,9 +863,9 @@ DValue *DtoCallFunction(const Loc &loc, Type *resulttype, DValue *fnval,
   }
 
   // get callee llvm value
-  LLValue *callable = DtoCallableValue(fnval);
-  LLFunctionType *const callableTy =
-      DtoExtractFunctionType(callable->getType());
+  LLValue *callable;
+  LLFunctionType *callableTy;
```
Comment (kinke) [inline gen/tocall.cpp:867]: Hmm, `callableTy = DtoType(tf)` might suffice here.

File: gen/tocall.cpp [hunk 2]
```diff
@@ -72,41 +72,30 @@ TypeFunction *DtoTypeFunction(DValue *fnval) {
...
+    v = DtoRVal(fn);
+    f = DtoType(fn->type);
   }
   if (type->ty == TY::Tdelegate) {
     if (fn->isLVal()) {
       LLValue *dg = DtoLVal(fn);
       LLValue *funcptr = DtoGEP(dg, 0, 1);
-      return DtoLoad(funcptr, ".funcptr");
+      v = DtoLoad(funcptr, ".funcptr");
+      f = DtoType(type->nextOf());
```
Comment (kinke) [inline gen/tocall.cpp:86]: missing early return

Comment (thewilsonator) [reply gen/tocall.cpp:86]: I'll try this, see if it works, then try `callableTy = DtoType(tf)`, if that works use it if not revert back.

File: gen/tocall.cpp (hunk 1 above)
Comment (thewilsonator) [reply gen/tocall.cpp:868]: Actually on second thoughts, for a delegate that would return the delegate type, not the underlying type of the function pointer of the delegate.

Comment (kinke) [reply gen/tocall.cpp:870]: ? It's already a `TypeFunction*`, so delegates are already handled by https://github.com/ldc-developers/ldc/blob/f142ac2e830b5a2865d8f5be8c34e5e2cd866069/gen/tocall.cpp#L46-L71.

Comment (thewilsonator) [reply gen/tocall.cpp]: OK trying this

Comment (thewilsonator) [reply gen/tocall.cpp]: `callableTy = isaFunction(DtoType(tf));` does not appear to work.

Comment (thewilsonator) [reply gen/tocall.cpp]: trying `callableTy = irFty.funcType;`

**Comment by thewilsonator** [comment]
OK this mostly works, except for `>>> TARGET FAILED: runnable/test22592.c`

**Comment by thewilsonator** [comment]
So the problem seems to be that 
```c
int testfn(void);
...
int testfn() { ... }
``` 
is treated as 
```llvm
call i32 (...) @testfn()
```
instead of 
```llvm
call i32 @testfn()
```
(I'm guessing)

**Comment by kinke** [comment]
The mismatch is in the frontend IIRC. This changed recently due K&R style functions support for importC; IIRC (fuzzy), a non-variadic fwd declaration can later get 'overwritten' by a variadic definition. See https://github.com/ldc-developers/ldc/blob/bd3d562d2a2eb41c2b8880473e3fb58484a471b6/gen/functions.cpp#L662-L668.

**Comment by thewilsonator** [comment]
Thanks, I'll see what I can make of it tomorrow.

File: gen/tocall.cpp [hunk 3]
```diff
@@ -877,10 +863,14 @@ DValue *DtoCallFunction(const Loc &loc, Type *resulttype, DValue *fnval,
   }
 
   // get callee llvm value
-  LLValue *callable = DtoCallableValue(fnval);
-  LLFunctionType *const callableTy =
-      DtoExtractFunctionType(callable->getType());
-  assert(callableTy);
+  LLValue *callable = DtoCallableValue(irFty.funcType, fnval);
+  LLFunctionType *callableTy = irFty.funcType;
+  if (dfnval && dfnval->func->isCsymbol()) {
```
Comment (thewilsonator) [inline gen/tocall.cpp:868]: this could made probably quite a bit faster with a function signature check, will get around to that soon.

**Pull request merged**